File: mockserver-client-dotnet/src/MockServer.Client/Models/Llm/LlmMockBuilder.cs
using MockServer.Client.Models;

namespace MockServer.Client.Llm;

/// <summary>
/// Fluent builder for a single LLM mock expectation (mirrors
/// <c>org.mockserver.client.LlmMockBuilder</c>). Builds one expectation whose
/// request matcher is <c>POST &lt;path&gt;</c> and whose action is an
/// <c>httpLlmResponse</c> carrying either a completion or an embedding.
/// </summary>
public sealed class LlmMockBuilder
{
    private readonly string _path;
    private string? _provider;
    private string? _model;
    private Completion? _completion;
    private EmbeddingResponse? _embedding;

    public LlmMockBuilder(string path)
    {
        _path = path;
    }

    /// <summary>Entry point mirroring <c>LlmMockBuilder.llmMock(path)</c>.</summary>
    public static LlmMockBuilder LlmMock(string path) => new(path);

    public LlmMockBuilder WithProvider(string provider)
    {
        _provider = provider;
        return this;
    }

    public LlmMockBuilder WithModel(string model)
    {
        _model = model;
        return this;
    }

    /// <summary>Responds with a completion (clears any previously set embedding).</summary>
    public LlmMockBuilder RespondingWith(Completion completion)
    {
        _completion = completion;
        _embedding = null;
        return this;
    }

    /// <summary>Responds with an embedding (clears any previously set completion).</summary>
    public LlmMockBuilder RespondingWith(EmbeddingResponse embedding)
    {
        _embedding = embedding;
        _completion = null;
        return this;
    }

    public Expectation Build()
    {
        var action = new HttpLlmResponse { Provider = _provider, Model = _model };
        if (_completion != null)
            action.Completion = _completion;
        if (_embedding != null)
            action.Embedding = _embedding;
        return new Expectation
        {
            HttpRequest = new HttpRequest { Method = "POST", Path = _path },
            HttpLlmResponse = action
        };
    }

    public List<Expectation> ApplyTo(MockServerClient client) => client.Upsert(Build());
}

File: mockserver-client-php/src/Llm/LlmMockBuilder.php
<?php

declare(strict_types=1);

namespace MockServer\Llm;

use MockServer\Expectation;
use MockServer\HttpRequest;
use MockServer\MockServerClient;

/**
 * Fluent builder for a single LLM completion or embedding mock
 * (mirrors org.mockserver.client.LlmMockBuilder).
 *
 * @example
 *   LlmMockBuilder::llmMock('/v1/chat/completions')
 *       ->withProvider(Provider::OPENAI)
 *       ->withModel('gpt-4o')
 *       ->respondingWith(Completion::completion()->withText('Hello!'))
 *       ->applyTo($client);
 */
class LlmMockBuilder
{
    private string $path;
    private ?string $provider = null;
    private ?string $model = null;
    private ?Completion $completion = null;
    private ?EmbeddingResponse $embedding = null;

    public function __construct(string $path)
    {
        $this->path = $path;
    }

    /**
     * Entry point mirroring {@code LlmMockBuilder.llmMock(path)}.
     */
    public static function llmMock(string $path): self
    {
        return new self($path);
    }

    public function withProvider(string $provider): self
    {
        $this->provider = $provider;
        return $this;
    }

    public function withModel(string $model): self
    {
        $this->model = $model;
        return $this;
    }

    /**
     * Respond with a completion or an embedding. Setting one clears the other.
     */
    public function respondingWith(Completion|EmbeddingResponse $response): self
    {
        if ($response instanceof EmbeddingResponse) {
            $this->embedding = $response;
            $this->completion = null;
        } else {
            $this->completion = $response;
            $this->embedding = null;
        }
        return $this;
    }

    public function build(): Expectation
    {
        $action = HttpLlmResponse::llmResponse();
        if ($this->provider !== null) {
            $action->withProvider($this->provider);
        }
        if ($this->model !== null) {
            $action->withModel($this->model);
        }
        if ($this->completion !== null) {
            $action->withCompletion($this->completion);
        }
        if ($this->embedding !== null) {
            $action->withEmbedding($this->embedding);
        }

        return (new Expectation())
            ->httpRequest(HttpRequest::request()->method('POST')->path($this->path))
            ->httpLlmResponse($action);
    }

    /**
     * Build the expectation and register it via the client.
     *
     * @return array<mixed>
     */
    public function applyTo(MockServerClient $client): array
    {
        return $client->upsertExpectation($this->build());
    }
}
